### ISSUE TITLE
Image size

### ISSUE THREAD

**Issue by user1103** [AlexeyAB/darknet#33]
@AlexeyAB if I use 416x416 in my config file, do I have to resize all my images to 416x416? If not, does darknet take in images of varying sizes or aren't square?

**Comment by VanitarNordic** [comment]
no, you don't need to do but keep your training image sizes bigger than 416*416. doesn't matter square or rectangle or whatever

**Comment by jng075** [comment]
@VanitarNordic does having it smaller than 416*416 affect mAP?

**Comment by VanitarNordic** [comment]
Yes, in my case it did affect, specially if the quantity is significant

**Comment by jng075** [comment]
I see, how did you calculate your mAP? If I recall correctly, no code does that, at least not in the original version of darknet.

**Comment by VanitarNordic** [comment]
Yes, there is no output as mAP but you can get the feeling when you consider avg errors and recalls for each iteration. maybe the author will add that later on

**Comment by jng075** [comment]
Yes I hope so too. I kept wondering how people were able to replicate their results if they didn't have the mAP calculations.

**Comment by jng075** [comment]
@VanitarNordic @AlexeyAB how would you try to improve mAP? I have 10,000 images of one class only, and I train with batch size 4, iterations 55,000. But results are quite poor and I need to change threshold from 0.001 to 0.2 (under detector recall function) to have some decent but still poor results.

I already tried to add a few more layers into cfg file. Would appreciate your help.

**Comment by VanitarNordic** [comment]
I am absolutely sure that you have made a mistake somewhere.
I trained YOLO-2 with only 100 images and 2000 iterations but the result was very good!

have you modified cfg file? have you renamed it? have you renamed the images containing folder? are you sure about your annotations? no need to add layers. actually with 10,000 images you should get a pretty decent detector, specially with one c;lass. make sure your validation set has enough number of images. I personally set to at least 1/4 of training images or even half.
Please do not modify anything except the ones mentioned in the readme.

**Comment by jng075** [comment]
@VanitarNordic I had to modify batch size because of memory issues. No renaming. I did not modify cfg file except to change number of classes, and the filter for last layer to (classes + coord + 1)*num. My validation has 2500 images, i.e. 7500 for training. I checked everything, my image folder, annotations are all correct. Inside the my-class.data cfg file (similar to voc.data) you can specify where your image folder is and I checked it.

I don't know if I am getting bad results because my images are totally different. I am using the darknet19_448.conv.23 weights. I even trained for 55,000 iterations. I wrote some code to calculate mAP and the results are not good. Recall values can be very high if I leave the thresholds the same, thresh = 0.001, iou_thresh = 0.5, and nms = 0.4 but mAP will be bad. Like 20-30% bad.

**Comment by VanitarNordic** [comment]
You must renamed the `.cfg` file to the mentioned one in the description, it is important that's reason he said you should renamed it. if you don't rename, the train and test work, but shows weird results, Do not touch the threshold at training please.

apart from your calculated mAP, you should get very good detection, is that correct with you or you get false-positives and false-negatives?

Could you please share your code till I test it with my own model?

**Comment by jng075** [comment]
@VanitarNordic you mean yolo-voc.cfg --> yolo-obj.cfg? I did that. Do you mind if I email you instead? This issues page is going out of topic now.

**Comment by VanitarNordic** [comment]
yes, you gonna do it because in my case I did not and for few hours I was scratching my head to find the problem, besides the folder which holds your training/test images should have the name which is written in readme. one time I change that to images and I was getting weird errors.

no problem if it goes out of topic, we gonna learn something from each other. The images in your validation set must be close to the images in real world which you decide to detect. it is logical, like you gonna train a kid to do something.

Yes, please. send you mAP calculation code. hesam dot moshiri at gmail dot com

**Comment by jng075** [comment]
@VanitarNordic please check your email

**Comment by AlexeyAB** [comment]
1. **You should not resize any images.** Yolo resizes images automatically.
2. **Yolo takes in aren't square images.** Yolo resizes it to square 416x416 image and then can detect objects on it with different aspect ratio, because trained on different aspect ratio. During training cut random rectangles with different aspect ratio - it allows to detect objects with aspect ratio distortion.

--------
In detail:

1. Yolo automatically resizes any images to network size 416x416 both for training and detection: https://github.com/AlexeyAB/darknet/blob/b3a3e92e8a482cf1c49322561077e8f3be54d619/src/detector.c#L468

2. During training Yolo cuts random rectangles with different aspect ratio from your training dataset images, it allows to detect objects with aspect ratio distortion:
    * random (left, right, top, bottom) for crop: https://github.com/AlexeyAB/darknet/blob/b3a3e92e8a482cf1c49322561077e8f3be54d619/src/data.c#L516
    * random crop: https://github.com/AlexeyAB/darknet/blob/b3a3e92e8a482cf1c49322561077e8f3be54d619/src/data.c#L528
    * random flip: https://github.com/AlexeyAB/darknet/blob/b3a3e92e8a482cf1c49322561077e8f3be54d619/src/data.c#L534
    * random changed colors: https://github.com/AlexeyAB/darknet/blob/b3a3e92e8a482cf1c49322561077e8f3be54d619/src/data.c#L535

**Comment by melaanya** [comment]
hello, @user1103! 
I'm also very interested in this problem with MaP calculating, don't yo mind sharing your code about MaP calculating with me, please?
